### ISSUE TITLE
BinaryDelta doesn't work in Mac monterey 

### ISSUE THREAD

**Issue by ZhuoxingGuo** [sparkle-project/Sparkle#1899]
when I use BinaryDelta to create patch in Mac monterey beta3 ,it doesn't work.

./BinaryDelta create --verbose Meeting.app Meeting2.app test.delta

it seems xar_prop_serializable failed .

![image](https://user-images.githubusercontent.com/6958119/126288059-d29cc0ba-7822-4756-a9e6-28e4ee61c10a.png)

when I use BinaryDelta to apply patch , it also fails in xar_open Method

<img width="1609" alt="wecom-temp-f205cc1dd1df780043173f2861075dcd" src="https://user-images.githubusercontent.com/6958119/126288935-5f9ffa3c-7d9f-477a-8893-36e4b0bd46a7.png">

 Can anyone help me ? Thanks

**Comment by zorgiepoo** [comment]
Can you submit a Feedback Assistant bug report to Apple and post a feedback number?

**Comment by ZhuoxingGuo** [comment]
> Can you submit a Feedback Assistant bug report to Apple and post a feedback number?

I have submited this issue to Apple, the feedback number  FB9373020 .

when I use command xar in macOS big sur or in macOS monterey,it says xar is no longer under active delvelopment by Apple .

Does it affect  incremental updates in the furture?  


<img width="667" alt="wecom-temp-6c50076e73e35520d7e0a461ec4303a6" src="https://user-images.githubusercontent.com/6958119/126431641-7a109ea0-b379-4d60-a886-525435bd78df.png">

**Comment by kornelski** [comment]
Should we migrate away from xar?

It may be an opportunity to develop a better binary delta format, or adopt an existing one.

Our current delta format is not great. Compression is less than what I'd expect. It doesn't handle files moved efficiently.

**Comment by zorgiepoo** [comment]
xar may not be under active development but it's used by Apple with things like xip (for Xcode builds) and flat pkg files. However we might be the only ones using it outside.. Once a long long time ago, Sparkle's binary delta support was a feature contributed by a WebKit engineer (at the time) so they could have their nightly builds update efficiently, I believe.

I don't think we should do any major undertaking in a Soon (2.0) release. We may need to fix our usage of xar here, maybe investigate but wait a bit closer to Monterey's release since it's a recent regression in a beta release. This could potentially involve a major version bump to our format too.

Also my impression is that delta patching isn't all that mature and mainstream and we are one of the main widespread users. There were a few bugs I had to investigate and fix in bsdiff itself a few years ago if I recall correctly, a couple of which surfaced due to our user base.

Maybe we could migrate to say libarchive at some point but I don't know if just the archive format could give us that much better compression (and has implications like do we use the OS version which gets updated, or bundle a non-trivial dependency ourselves with a larger security surface..)

**Comment by kornelski** [comment]
I'm worried that xar is really dead, and the API fails because Apple has decided to kill it.

I think we could drop xar and bsdiff and replace them with zstd and some basic container format.

zstd, besides having decent compression and great speed, has an option to specify a "prefix" for compression/decompression. This prefix is automatically used as an optimal file diff. 

Prefix is a one-time dictionary. zstd also has an option to generate reusable dictionaries, but AFAIK this isn't suitable for our case, since the dictionaries are slow to generate and are supposed to be reusable for many compressions/decompressions.

bsdiff has some basic smarts for dealing with changed offsets in executables (although that's closer to PNG filters than a disassembler). I'll test if zstd's prefix is efficient enough to beat it.

**Comment by zorgiepoo** [comment]
I would still hold off until on making a hasty decision until I’ve time to make a proper assessment/investigation here on this issue for the near future.

**Comment by zorgiepoo** [comment]
On even Big Sur (macOS 11.5):

```
~/D/P/Sparkle (2.x)> xar -t -f /Users/me/Downloads/1137_to_1139.delta 
Warning, archive contains invalid path: /Contents/CodeResources
Warning, archive contains invalid path: /Contents/Frameworks/Sparkle.framework/Autoupdate
...bunch more warnings...
```
The issue is Sparkle is not storing files in the container properly. It expects a hierarchal structure of nodes, like other formats do probably. I should be able to fix this fairly quickly; it shouldn't need a major version bump and won't need to break compatibility with applying new patches in older Sparkle versions.

We should still switch to a new format given the deprecation warning however. Let's track that as a separate issue.

**Comment by ZhuoxingGuo** [comment]
Anyway, We and my compay are very appreciate what your guys done , you have made a greate achivement in Mac App upgrading. Thanks a lot 。

**Comment by zorgiepoo** [comment]
BinaryDelta support should be fixed now in top of tree. The delta files may need to be re-generated to work on that beta version of Monterey; when Monterey is actually released we may get a better idea of implications here.

**Comment by zorgiepoo** [comment]
The story for BinaryDelta now that Monterey (and updates to Big Sur / Catalina) are out are something like the follows:

* Older versions of Sparkle (pre-1.27.1) will still be able to patch old patches created prior to 1.27.0 even on Monterey / recent OS's (contrary to this report that exhibited issues in an earlier Monterey beta build)
* When *creating* patches, you must use Sparkle 1.27.0 or later on new/recent OS updates, otherwise they won't work.
* If you use Sparkle 1.27.1 you now *cannot* apply patches that were created with older Sparkle relases (pre 1.27.0)
* New patches created with Sparkle 1.27.0 or later can still be applied from older versions of Sparkle (pre 1.27.0).

The long story short is, when creating new patches, use Sparkle 1.27.0 or later and you'll be fine.

**Comment by C4Framework** [comment]
@zorgiepoo I am getting the `xar_prop_serializable failed` error above

I'm on a new machine running Monterey, and Sparkle 1.27.1.

Previous versions of my app were generated with Sparkle 1.25

What can I do to resolve this?

**Comment by zorgiepoo** [comment]
Are you using generate_appcast (or BinaryDelta) tools from 1.27.1 (the tools are the important bit, not the framework itself)?

If so, I may need old and new binary to investigate and try to create a patch myself..

**Comment by C4Framework** [comment]
We're using `generate_appcast`.

Also, apologies if I misread the ticket above...

Here's the last version of our app, which was created on Catalina with Sparkle 1.25:
[Flow 1.12.2](https://createwithflow-releases.s3-us-west-2.amazonaws.com/Flow-1122.dmg)

Here's the newest version, compiled on Monterey / Silicon, with Sparkle 1.27:
[Flow 1.13](https://createwithflow-releases.s3-us-west-2.amazonaws.com/Flow.dmg)

**I'd also like to note that it appears that despite the failure / warnings, updating from 1.12.2 to 1.13.0 seems to actually work.**

**Comment by zorgiepoo** [comment]
To clarify again, the `generate_appcast` binary must be from our Sparkle 1.27 (or later) distribution. You can't use an old version of generate_appcast from our 1.25 distribution. Please check this!

I can generate and apply a patch fine on my Monterey machine using a BinaryDelta from Sparkle 1.27.1.

> I'd also like to note that it appears that despite the failure / warnings, updating from 1.12.2 to 1.13.0 seems to actually work.

Sparkle falls back on regular update archive if binary delta patching fails. Also this issue is to do with *creating* valid patches on recent OS's. Applying older patches may still work from older Sparkle versions.

**Comment by zorgiepoo** [comment]
We have a new delta update format. Beta support landed in [2.1.0-beta.1](https://github.com/sparkle-project/Sparkle/releases/tag/2.1.0-beta.1) to try out.
